File: src/app/interact.zig
//! Pure helpers for interactive terminal actions: token extraction at a column,
//! line-suffix stripping, and URL-vs-path classification. No platform I/O.
//! These are extracted for unit testability; the glue lives in main.zig.

const std = @import("std");

/// Extract the whitespace-delimited token from `line` that spans column `col`.
/// `line` is a UTF-8 string; `col` is a byte-column index (0-based).
/// Returns a sub-slice of `line`, or empty if `col` is out of range or the
/// character at `col` is whitespace.
pub fn tokenAtCol(line: []const u8, col: usize) []const u8 {
    if (col >= line.len) return "";
    // If the clicked character is whitespace, there's no token here.
    if (std.ascii.isWhitespace(line[col])) return "";

    // Scan left to find the start of the token.
    var start = col;
    while (start > 0 and !std.ascii.isWhitespace(line[start - 1])) {
        start -= 1;
    }

    // Scan right to find the end of the token.
    var end = col + 1;
    while (end < line.len and !std.ascii.isWhitespace(line[end])) {
        end += 1;
    }

    return line[start..end];
}

/// Strip a trailing `:line` or `:line:col` numeric suffix from `tok`.
/// E.g. "src/main.zig:412" → "src/main.zig", "foo.zig:10:3" → "foo.zig".
/// Returns a sub-slice of `tok` (no allocation).
pub fn stripLineSuffix(tok: []const u8) []const u8 {
    // Work backwards: strip an optional `:digits` suffix up to twice.
    var s = tok;

    // Try to strip `:col` first (second suffix), then `:line` (first suffix).
    // We do this twice so we handle both `:line` and `:line:col`.
    var iterations: usize = 0;
    while (iterations < 2) : (iterations += 1) {
        // Find the last ':' in s.
        const colon = std.mem.lastIndexOfScalar(u8, s, ':') orelse break;
        if (colon == 0) break; // colon at start — no path before it
        const after = s[colon + 1 ..];
        if (after.len == 0) break; // trailing colon with nothing after
        // Verify all chars after ':' are ASCII digits.
        var all_digits = true;
        for (after) |ch| {
            if (ch < '0' or ch > '9') {
                all_digits = false;
                break;
            }
        }
        if (!all_digits) break;
        s = s[0..colon];
    }
    return s;
}

/// Classification of a terminal token for ⌘-click handling.
pub const Kind = enum { url, path, none };

/// Classify `tok` (already suffix-stripped) as a URL, a file path, or neither.
/// `cwd` is used to probe for file existence when other heuristics are
/// inconclusive. Pass an empty slice to skip the existence check.
pub fn classify(tok: []const u8, cwd: []const u8) Kind {
    if (tok.len == 0) return .none;

    // URL: starts with http:// or https://.
    if (std.mem.startsWith(u8, tok, "http://") or
        std.mem.startsWith(u8, tok, "https://")) return .url;

    // Absolute path.
    if (tok[0] == '/') return .path;

    // Relative path heuristics: contains '/' or has a '.' extension.
    const has_slash = std.mem.indexOfScalar(u8, tok, '/') != null;
    const has_ext = blk: {
        // A dot that is not the first character and has at least one char after it.
        if (std.mem.lastIndexOfScalar(u8, tok, '.')) |dot| {
            break :blk dot > 0 and dot + 1 < tok.len;
        }
        break :blk false;
    };

    if (has_slash or has_ext) return .path;

    // Last resort: does the file exist relative to cwd?
    if (cwd.len > 0 and fileExistsRelative(cwd, tok)) return .path;

    return .none;
}

/// True if `name` exists (any type) relative to `dir`. Uses POSIX `access(2)`.
fn fileExistsRelative(dir: []const u8, name: []const u8) bool {
    // Build a null-terminated path: dir + "/" + name.
    var buf: [std.fs.max_path_bytes + 1]u8 = undefined;
    const path = std.fmt.bufPrint(buf[0 .. buf.len - 1], "{s}/{s}", .{ dir, name }) catch return false;
    buf[path.len] = 0;
    const path_z: [*:0]const u8 = buf[0..path.len :0];
    return std.c.access(path_z, 0) == 0; // F_OK = 0
}

// --- Tests ------------------------------------------------------------------

const testing = std.testing;

test "tokenAtCol: basic token extraction" {
    try testing.expectEqualStrings("hello", tokenAtCol("hello world", 0));
    try testing.expectEqualStrings("hello", tokenAtCol("hello world", 3));
    try testing.expectEqualStrings("hello", tokenAtCol("hello world", 4));
    try testing.expectEqualStrings("world", tokenAtCol("hello world", 6));
    try testing.expectEqualStrings("world", tokenAtCol("hello world", 10));
}

test "tokenAtCol: whitespace at col returns empty" {
    try testing.expectEqualStrings("", tokenAtCol("hello world", 5));
}

test "tokenAtCol: col out of range returns empty" {
    try testing.expectEqualStrings("", tokenAtCol("hi", 99));
}

test "tokenAtCol: single token" {
    try testing.expectEqualStrings("src/main.zig:412", tokenAtCol("src/main.zig:412", 5));
}

test "tokenAtCol: leading whitespace" {
    try testing.expectEqualStrings("foo", tokenAtCol("  foo  bar", 3));
    try testing.expectEqualStrings("bar", tokenAtCol("  foo  bar", 7));
}

test "stripLineSuffix: strips :line" {
    try testing.expectEqualStrings("src/main.zig", stripLineSuffix("src/main.zig:412"));
}

test "stripLineSuffix: strips :line:col" {
    try testing.expectEqualStrings("src/main.zig", stripLineSuffix("src/main.zig:412:3"));
}

test "stripLineSuffix: no suffix unchanged" {
    try testing.expectEqualStrings("src/main.zig", stripLineSuffix("src/main.zig"));
}

test "stripLineSuffix: non-digit suffix unchanged" {
    try testing.expectEqualStrings("foo:bar", stripLineSuffix("foo:bar"));
}

test "stripLineSuffix: trailing colon unchanged" {
    try testing.expectEqualStrings("foo:", stripLineSuffix("foo:"));
}

test "classify: URLs" {
    try testing.expectEqual(Kind.url, classify("http://example.com", ""));
    try testing.expectEqual(Kind.url, classify("https://example.com/path", ""));
}

test "classify: absolute path" {
    try testing.expectEqual(Kind.path, classify("/usr/local/bin", ""));
}

test "classify: relative path with slash" {
    try testing.expectEqual(Kind.path, classify("src/main.zig", ""));
}

test "classify: relative path with extension" {
    try testing.expectEqual(Kind.path, classify("main.zig", ""));
}

test "classify: bare word with no heuristics" {
    try testing.expectEqual(Kind.none, classify("hello", ""));
}

test "classify: empty token" {
    try testing.expectEqual(Kind.none, classify("", ""));
}

File: crates/anvil-workspace/src/interact.rs
//! Pure helpers for interactive terminal actions: token extraction at a column,
//! line-suffix stripping, and URL-vs-path classification.  No platform I/O
//! beyond an optional `access(2)` existence check.

use std::path::Path;

/// Extract the whitespace-delimited token from `line` that spans byte column
/// `col`.  Returns a sub-slice of `line`, or `""` if `col` is out of range or
/// the character at `col` is ASCII whitespace.
pub fn token_at_col(line: &str, col: usize) -> &str {
    let bytes = line.as_bytes();
    if col >= bytes.len() {
        return "";
    }
    if bytes[col].is_ascii_whitespace() {
        return "";
    }

    // Scan left for start.
    let mut start = col;
    while start > 0 && !bytes[start - 1].is_ascii_whitespace() {
        start -= 1;
    }

    // Scan right for end.
    let mut end = col + 1;
    while end < bytes.len() && !bytes[end].is_ascii_whitespace() {
        end += 1;
    }

    &line[start..end]
}

/// Strip a trailing `:line` or `:line:col` numeric suffix from `tok`.
/// E.g. `"src/main.zig:412"` → `"src/main.zig"`.
/// Returns a sub-slice of `tok` (no allocation).
pub fn strip_line_suffix(tok: &str) -> &str {
    let mut s = tok;
    for _ in 0..2 {
        let colon = match s.rfind(':') {
            Some(i) if i > 0 => i,
            _ => break,
        };
        let after = &s[colon + 1..];
        if after.is_empty() {
            break;
        }
        if !after.bytes().all(|b| b.is_ascii_digit()) {
            break;
        }
        s = &s[..colon];
    }
    s
}

/// Classification of a terminal token for ⌘-click handling.
#[derive(Clone, PartialEq, Eq, Debug)]
pub enum Kind {
    Url,
    Path,
    /// A path with a trailing `:line` or `:line:col` suffix — the click
    /// handler should open the file at that location instead of the top.
    PathWithLine {
        path: String,
        line: u32,
        col: Option<u32>,
    },
    None,
}

/// Parse a `foo.rs:42` or `foo.rs:42:7` token into its parts.
///
/// Rules:
///   - The suffix must be all-digit segments separated by colons.
///   - Line is required and must be > 0 (editors are 1-based).
///   - Col is optional. A trailing `:0` for col is allowed and surfaced.
///   - Returns `None` if no recognizable line suffix.
pub fn parse_path_with_line(tok: &str) -> Option<(String, u32, Option<u32>)> {
    // Right-to-left: peel up to two `:N` segments. A non-numeric tail or
    // missing colon stops the scan (preserves whatever's already been
    // collected); we don't bail because the path itself may legitimately
    // contain colons earlier (e.g. `/some/path:weird`).
    let mut s = tok;
    let mut nums: Vec<u32> = Vec::new();
    for _ in 0..2 {
        let Some(colon) = s.rfind(':') else {
            break;
        };
        if colon == 0 {
            break;
        }
        let after = &s[colon + 1..];
        if after.is_empty() || !after.bytes().all(|b| b.is_ascii_digit()) {
            break;
        }
        let Ok(n) = after.parse() else {
            break;
        };
        nums.push(n);
        s = &s[..colon];
    }
    let path = s.to_string();
    match nums.as_slice() {
        [line] if *line > 0 => Some((path, *line, None)),
        [col, line] if *line > 0 => Some((path, *line, Some(*col))),
        _ => None,
    }
}

/// Classify `tok` (raw, may contain `:line[:col]` suffix) as a URL, a path
/// with optional line, a plain path, or neither.
///
/// `cwd` is used to probe for file existence when other heuristics are
/// inconclusive.  Pass `""` to skip the existence check.
pub fn classify(tok: &str, cwd: &str) -> Kind {
    if tok.is_empty() {
        return Kind::None;
    }

    if tok.starts_with("http://") || tok.starts_with("https://") {
        return Kind::Url;
    }

    // Path with `:line[:col]` suffix → upgrade Path → PathWithLine when the
    // base classifies as a path.
    if let Some((path, line, col)) = parse_path_with_line(tok) {
        if classify_plain(&path, cwd) == Kind::Path {
            return Kind::PathWithLine { path, line, col };
        }
    }

    classify_plain(tok, cwd)
}

/// Classify without considering the `:line[:col]` suffix. Kept separate so
/// `classify` can re-use it for both the suffixed and bare-path paths.
fn classify_plain(tok: &str, cwd: &str) -> Kind {
    if tok.starts_with('/') {
        return Kind::Path;
    }

    let has_slash = tok.contains('/');
    let has_ext = tok
        .rfind('.')
        .map(|dot| dot > 0 && dot + 1 < tok.len())
        .unwrap_or(false);

    if has_slash || has_ext {
        return Kind::Path;
    }

    if !cwd.is_empty() && file_exists_relative(cwd, tok) {
        return Kind::Path;
    }

    Kind::None
}

fn file_exists_relative(dir: &str, name: &str) -> bool {
    Path::new(dir).join(name).exists()
}

// ---------------------------------------------------------------------------
// Tests  (13 Zig tests → 13 Rust tests)
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn token_at_col_basic_extraction() {
        assert_eq!(token_at_col("hello world", 0), "hello");
        assert_eq!(token_at_col("hello world", 3), "hello");
        assert_eq!(token_at_col("hello world", 4), "hello");
        assert_eq!(token_at_col("hello world", 6), "world");
        assert_eq!(token_at_col("hello world", 10), "world");
    }

    #[test]
    fn token_at_col_whitespace_at_col_returns_empty() {
        assert_eq!(token_at_col("hello world", 5), "");
    }

    #[test]
    fn token_at_col_col_out_of_range_returns_empty() {
        assert_eq!(token_at_col("hi", 99), "");
    }

    #[test]
    fn token_at_col_single_token() {
        assert_eq!(token_at_col("src/main.zig:412", 5), "src/main.zig:412");
    }

    #[test]
    fn token_at_col_leading_whitespace() {
        assert_eq!(token_at_col("  foo  bar", 3), "foo");
        assert_eq!(token_at_col("  foo  bar", 7), "bar");
    }

    #[test]
    fn strip_line_suffix_strips_line() {
        assert_eq!(strip_line_suffix("src/main.zig:412"), "src/main.zig");
    }

    #[test]
    fn strip_line_suffix_strips_line_col() {
        assert_eq!(strip_line_suffix("src/main.zig:412:3"), "src/main.zig");
    }

    #[test]
    fn strip_line_suffix_no_suffix_unchanged() {
        assert_eq!(strip_line_suffix("src/main.zig"), "src/main.zig");
    }

    #[test]
    fn strip_line_suffix_non_digit_suffix_unchanged() {
        assert_eq!(strip_line_suffix("foo:bar"), "foo:bar");
    }

    #[test]
    fn strip_line_suffix_trailing_colon_unchanged() {
        assert_eq!(strip_line_suffix("foo:"), "foo:");
    }

    #[test]
    fn classify_urls() {
        assert_eq!(classify("http://example.com", ""), Kind::Url);
        assert_eq!(classify("https://example.com/path", ""), Kind::Url);
    }

    #[test]
    fn classify_absolute_path() {
        assert_eq!(classify("/usr/local/bin", ""), Kind::Path);
    }

    #[test]
    fn classify_relative_path_with_slash() {
        assert_eq!(classify("src/main.zig", ""), Kind::Path);
    }

    #[test]
    fn classify_relative_path_with_extension() {
        assert_eq!(classify("main.zig", ""), Kind::Path);
    }

    #[test]
    fn classify_bare_word_no_heuristics() {
        assert_eq!(classify("hello", ""), Kind::None);
    }

    #[test]
    fn classify_empty_token() {
        assert_eq!(classify("", ""), Kind::None);
    }

    #[test]
    fn classify_bare_word_file_exists_relative_to_cwd() {
        // Create a temp file without any slash or extension so heuristics
        // won't classify it as a path, but file_exists_relative will.
        let dir = std::env::temp_dir().join("anvil_interact_test");
        let _ = std::fs::create_dir_all(&dir);
        let fname = "Makefile"; // no slash, no extension
        let fpath = dir.join(fname);
        std::fs::write(&fpath, b"").expect("write tmp file");
        let cwd = dir.to_string_lossy().into_owned();
        assert_eq!(classify(fname, &cwd), Kind::Path);
        let _ = std::fs::remove_file(&fpath);
        let _ = std::fs::remove_dir(&dir);
    }

    #[test]
    fn classify_bare_word_file_not_found_returns_none() {
        // A bare word that doesn't exist on disk with a valid cwd → Kind::None.
        assert_eq!(classify("nonexistent_bare_word_xyz", "/tmp"), Kind::None);
    }

    #[test]
    fn parse_path_with_line_recognizes_line_only() {
        let (p, l, c) = parse_path_with_line("foo.rs:42").unwrap();
        assert_eq!(p, "foo.rs");
        assert_eq!(l, 42);
        assert_eq!(c, None);
    }

    #[test]
    fn parse_path_with_line_recognizes_line_and_col() {
        let (p, l, c) = parse_path_with_line("foo.rs:42:7").unwrap();
        assert_eq!(p, "foo.rs");
        assert_eq!(l, 42);
        assert_eq!(c, Some(7));
    }

    #[test]
    fn parse_path_with_line_handles_absolute_path() {
        let (p, l, c) = parse_path_with_line("/abs/path.rs:10").unwrap();
        assert_eq!(p, "/abs/path.rs");
        assert_eq!(l, 10);
        assert_eq!(c, None);
    }

    #[test]
    fn parse_path_with_line_no_suffix_returns_none() {
        assert!(parse_path_with_line("foo.rs").is_none());
    }

    #[test]
    fn parse_path_with_line_zero_line_rejected() {
        // Editors are 1-based — 0 is meaningless.
        assert!(parse_path_with_line("foo.rs:0").is_none());
    }

    #[test]
    fn classify_path_with_line_returns_pathwithline() {
        match classify("src/main.rs:42:7", "") {
            Kind::PathWithLine { path, line, col } => {
                assert_eq!(path, "src/main.rs");
                assert_eq!(line, 42);
                assert_eq!(col, Some(7));
            }
            other => panic!("expected PathWithLine, got {other:?}"),
        }
    }
}
